### ISSUE TITLE
Unexpected? unfolding in simple program

### ISSUE THREAD

**Issue by omelkonian** [agda/agda#6130]
*Agda v2.6.2.1*

----------------------

Given the following predicate on `List.Any` proofs that point to the last element:
```agda
open import Agda.Builtin.List
open import Agda.Builtin.Equality

private variable
  A : Set; P : A → Set
  x : A; xs : List A

data Any (P : A → Set) : List A → Set where
  here : P x → Any P (x ∷ xs)
  there : Any P xs → Any P (x ∷ xs)

LastAny : Any P xs → Set
LastAny = λ where
  (here {xs = xs} _) → xs ≡ []
  (there p) → LastAny p
```
We can get Agda to produce two holes that have repeatedly unfolded:
```agda
test : (p : Any P xs) → LastAny p → LastAny p
-- ↓ works
-- test _ = λ p → p

-- ↓ works
-- test = λ where
--   (here _)  → λ p → p
--   (there _) → λ p → p

-- ↓ works
-- test = λ where
--   (here _)  → λ p → p
--   (there p) → test p

-- ↓ produces two huge goals:
test = λ where
  (here _)  → λ p → p
  (there _) → test {!!} {!!}
```
The content of the holes is:
```
Goal: Any _P_62
      (_x_71 ∷
       _x_79 ∷
       _x_87 ∷
       _x_95 ∷
       _x_103 ∷
       _x_111 ∷
       _x_119 ∷
       _x_127 ∷
       _x_135 ∷
       _x_143 ∷
       _x_151 ∷
       _x_159 ∷
       _x_167 ∷
       _x_175 ∷
       _x_183 ∷
       _x_191 ∷
       _x_199 ∷
       _x_207 ∷
       _x_215 ∷
       _x_223 ∷
       _x_231 ∷
       _x_239 ∷
       _x_247 ∷
       _x_255 ∷
       _x_263 ∷
       _x_271 ∷
       _x_279 ∷
       _x_287 ∷
       _x_295 ∷
       _x_303 ∷
       _x_311 ∷
       _x_319 ∷
       _x_327 ∷
       _x_335 ∷
       _x_343 ∷
       _x_351 ∷
       _x_359 ∷
       _x_367 ∷
       _x_375 ∷
       _x_383 ∷
       _x_391 ∷
       _x_399 ∷
       _x_407 ∷
       _x_415 ∷
       _x_423 ∷ _x_431 ∷ _x_439 ∷ _x_447 ∷ _x_455 ∷ _x_463 ∷ _xs_464)
————————————————————————————————————————————————————————————
xs₁  : List xs.A   (not in scope)
xs₁  = x₁ ∷ xs
x    : Any P xs   (not in scope)
x₁   : xs.A   (not in scope)
xs   : List xs.A   (not in scope)
P    : xs.A → Set   (not in scope)
xs.A : Set   (not in scope)
———— Constraints ———————————————————————————————————————————
_466
  := λ {xs} {x} x₁ →
       test
       (there
        (there
         (there
          (there
           (there
            (there
             (there
              (there
               (there
                (there
                 (there
                  (there
                   (there
                    (there
                     (there
                      (there
                       (there
                        (there
                         (there
                          (there
                           (there
                            (there
                             (there
                              (there
                               (there
                                (there
                                 (there
                                  (there
                                   (there
                                    (there
                                     (there
                                      (there
                                       (there
                                        (there
                                         (there
                                          (there
                                           (there
                                            (there
                                             (there
                                              (there
                                               (there
                                                (there
                                                 (there
                                                  (there
                                                   (there
                                                    (there
                                                     (there
                                                      (there
                                                       (there
                                                        (there
                                                         (_p_465
                                                          (x =
                                                           x₁))))))))))))))))))))))))))))))))))))))))))))))))))))
       (?1 (x = x₁))
  (blocked on problem 83)
```
Granted, the expression that leads to this is non-sense, but is this behaviour expected or does it point to some issue that might slow down type-checking?

**Comment by andreasabel** [comment]
The behavior can be explained by the warning given:
> Refusing to invert pattern matching of `.extendedlambda0` because the
> maximum depth (50) has been reached. Most likely this means you
> have an unsatisfiable constraint, but it could also mean that you
> need to increase the maximum depth using the flag
> `--inversion-max-depth=N`
> when checking that the expression `test ? ?` has type
> `LastAny (there x) → LastAny (there x)`

I am not familiar with this component of the Agda type checker.  
@UlfNorell , @jespercockx ?

**Comment by jespercockx** [comment]
Removing the unnecessary pattern lambdas and adding `--inversion-max-depth=3`, you get:
```agda
{-# OPTIONS --inversion-max-depth=3 #-}

open import Agda.Builtin.List
open import Agda.Builtin.Equality

private variable
  A : Set; P : A → Set
  x : A; xs : List A

data Any (P : A → Set) : List A → Set where
  here : P x → Any P (x ∷ xs)
  there : Any P xs → Any P (x ∷ xs)

LastAny : Any P xs → Set
LastAny (here {xs = xs} _) = xs ≡ []
LastAny (there p) = LastAny p

test : (p : Any P xs) → LastAny p → LastAny p
test (here _)  = λ p → p
test (there _) = test {!!} {!!}
```
Error:
```
———— Errors ————————————————————————————————————————————————
Failed to solve the following constraints:
  LastAny _p_89 =< LastAny (there x) → LastAny (there x)
    (blocked on _p_89)

———— Warnings ——————————————————————————————————————————————
Refusing to invert pattern matching of LastAny because the maximum
depth (3) has been reached. Most likely this means you have an
unsatisfiable constraint, but it could also mean that you need to
increase the maximum depth using the flag --inversion-max-depth=N
when checking that the expression test ? ? has type
LastAny (there x) → LastAny (there x)
```
Basically, Agda sees the constraint `LastAny _p_89 =< LastAny (there x) → LastAny (there x)` and is smart enough to see that the meta `_p_89` should **not** be a `here`, since that would produce the type `xs ≡ []` which is definitely not a function type. So the inversion machinery in the constraint solver assumes it has to be the other case, i.e. `_p_89` is some application of `there`. Unfortunately in this case, this leads nowhere either, since you again get (almost) the same problem.

Perhaps the function inverse machinery could be made smarter by checking whether there is *any* chance that `LastAny` reduces to a function type, and refuse to do the inversion if not. However, I am not sure whether this is feasible/desirable in general.

**Comment by andreasabel** [comment]
So, this is part of the "constructor-headed"/injectivity component, is it?  Couldn't we remember which constructors (including type constructors) a definition can produce, and only do the expansion if the rhs constructor is in this set?  ("rhs" relative to a constraint `f X = rhs`.)  We already check that all definition rhs are constructors, so we could store this set, can we?

**Comment by jespercockx** [comment]
Yes, that's correct. Remembering the set of possible head symbols is something we already do I believe, the tricky thing in this case is that one of the heads is not rigid but rather a recursive call. In general, this could be a call to a *different* injective function (or even a non-injective one), we just treat it as a fallback case for when none of the other head symbols match the one we are looking for.

**Comment by andreasabel** [comment]
So maybe the computation of the possible head symbols needs to be recursive?  Atm, it isn't:
https://github.com/agda/agda/blob/8a2ab7b7fc91b5dbbe49bb99838a9f28bd2528ec/src/full/Agda/TypeChecking/Injectivity.hs#L109
Here, we would need to fill in the possible head symbols of `f` (which would be the empty set when `f` is the function whose head symbols we are currently computing).  For mutual functions, this would have to be organized as a saturation algorithm.

P.S.: Thinking about it, maybe there need to be two calculations, the current one (that computes possible head symbols to encourage the expansion), and one that computes the possible head symbols that can arise after expansions (to discourage expansions like in the OP).
